tiledb/sm/rest/rest_client.cc

```diff
@@ -102,14 +102,17 @@ Status RestClient::get_array_schema_from_rest(
     const URI& uri, ArraySchema** array_schema) {
   // Init curl and form the URL
   Curl curlc;
-  RETURN_NOT_OK(curlc.init(config_, extra_headers_));
+  RETURN_NOT_OK(curlc.init(config_, extra_headers_, &redirect_meta_));
   std::string array_ns, array_uri;
   RETURN_NOT_OK(uri.get_rest_components(&array_ns, &array_uri));
-  std::string url = rest_server_ + "/v1/arrays/" + array_ns + "/" +
+  std::string url = (redirect_uri().empty() ? rest_server_ : redirect_uri()) +
+                    "/v1/arrays/" + array_ns + "/" +
                     curlc.url_escape(array_uri);
 
   // Get the data
   Buffer returned_data;
+  std::unique_lock<std::mutex> rd_lck(redirect_mtx_);
```

Comment:
we don't need to lock every time we make a request, we only need to lock access to `redirect_meta_`. Pass the mutex in `curlc.init` and pass it along side `redirect_meta_` into the callback in either a struct or a pair.